TST: Fix various incorrect linalg tests

eric-wieser opened the pull request:
The test cases for linalg supposedly include an empty test case. However, this does not actually test the right thing, since this case is supposed to be square, but is not.

This is because `atleast_2d(array([], dtype=double))` returns an array of shape (1, 0), not (0, 0)

Not only that, but when a test is marked as working for both square and non-square inputs, only the square inputs are run, because the non-square ones are shadowed by a copy-and-paste-o. As a result, the non-square inputs don't all even make sense, because they were never being run.

---

After fixing the tests, fixes a bunch of minor bugs exposed by the updated tests, where instead of `LinAlgError` being thrown, some  much-more-internal error was escaping, that hid the cause of the problem.

eric-wieser committed:
TST: Correct empty square test case to actually be square

atleast_2d(array([], dtype=double)) returns an array of shape (1, 0), not (0, 0)

eric-wieser committed:
TST: Add some non-square 0-shaped test-cases

eric-wieser committed:
TST: Correct pinv test case such that it doesn't fail correct cases

eric-wieser committed:
TST: Prevent non-square testcases being hidden by square ones (fix typo)

eric-wieser committed:
TST: Enable testing pinv on non-square matrices

eric-wieser commented:
Should `LinalgNonsquareTestCase` be a subclass of `LinalgTestCase`?

eric-wieser committed:
TST: Adjust the precision of assert_almost_equal, but based on the type

SVD was previously being too generous with accuracy on doubles. This allows
pinv, the test for which was previously was too imprecise, to pass.

eric-wieser committed:
TST: Refactor all the test case lists

Allows each individual function to inspect the flags of a certain test, and
decide whether an exception will be thrown

eric-wieser commented on numpy/linalg/tests/test_linalg.py:
```diff
@@ -338,7 +338,7 @@ def test_sq_cases(self):
 
 class LinalgNonsquareTestCase(object):
 
-    def test_sq_cases(self):
```
This meant that none of the non-square test cases were actually being run...

eric-wieser replied:
Now that they are run, they fail, because

```python
>>> import numpy as np
>>> a = np.random.rand(8, 11)
>>> b = np.random.rand(11)
>>> np.linalg.lstsq(a, b)
```

doesn't make any sense

charris replied:
Has been fixed.

eric-wieser committed:
TST: Correct test cases to actually make sense

These testcases were never used in the first place, due to a typo. This makes
their dimensions match the order of the other test cases, even though those
also did not run

eric-wieser commented:
Fingers crossed, every failure after this point should be a real one that was not being correctly tested for.

eric-wieser committed:
BUG: prevent np.linalg.eig ValueError when given a 0x0 array

Throw LinAlgError instead with an appropriate message

eric-wieser commented on numpy/linalg/linalg.py:
```diff
@@ -1933,36 +1934,40 @@ def lstsq(a, b, rcond=-1):
     a, bstar = _fastCopyAndTranspose(t, a, bstar)
     a, bstar = _to_native_byte_order(a, bstar)
     s = zeros((min(m, n),), real_t)
-    nlvl = max( 0, int( math.log( float(min(m, n))/2. ) ) + 1 )
```
This was meant to be `log2`, apparently, and that `2` was supposed to be `1 + magic_internal_lapack_number`. Better to let LAPACK calculate it, I think.

eric-wieser commented on numpy/linalg/linalg.py:
```diff
@@ -1947,6 +1947,12 @@ def lstsq(a, b, rcond=-1):
+            raise RuntimeError(
+                'LAPACK did not compute workspace sizes: len(work) = {}, len(rwork) = {}, len(iwork) = {}'
+                .format(lwork, lrwork, liwork)
+            )
```
This is a known and fixed (9 Jun 2010) bug in LAPACK: http://icl.cs.utk.edu/lapack-forum/archives/lapack/msg00899.html/.

However, this fails on Appveyor. What version of LAPACK is being used, and do we really need to support a version released before python 2.7.0?

eric-wieser commented on numpy/linalg/linalg.py:
```diff
@@ -1963,6 +1969,11 @@ def lstsq(a, b, rcond=-1):
+            raise RuntimeError(
+                'LAPACK did not compute workspace sizes: len(work) = {}, len(iwork) = {}'
+                .format(lwork, liwork)
+            )
```
This is a known and fixed (9 Jun 2010) bug in LAPACK: http://icl.cs.utk.edu/lapack-forum/archives/lapack/msg00899.html/. The `scipy.linalg` code has a very similar test, that decides to just error out if this bug is present in LAPACK.

This condition is not hit on my machine (Windows, Intel MKL).

However, this fails on Appveyor. What version of LAPACK is being used there, and do we really need to support a version released before python 2.7.0?

ilayn replied:
It turns out that Accelerate based problems creep up everywhere, I've recently bitten by the same lapack version problem -> Scipy discussion https://github.com/scipy/scipy/pull/6051

ilayn replied:
Oh, that's indeed prehistoric.

pv commented:
The description of this PR doesn't actually say what it does.
What was incorrect, and how does this address it?

eric-wieser commented:
@pv: I've updated the description. There was a little scope creep here - I'll aim to cut it back in the next day or two to remedy that.

eric-wieser committed:
BUG: Raise LinAlgError from lstsq rather than a math domain error from math.log

We could make the log conditional on its argument being non-zero, but that entire expression is wrong anyway.
We could omit that calculation entirely and have LAPACK calculate it for us, but the routine in LAPACK is wrong anyway
We could upgrade the version of lapack shipped in lapack_lite, but the tool to do that is wrong anyway.

Let's leave that can of worms for another time, and just improve the error message for now.

eric-wieser commented on numpy/linalg/linalg.py:
```diff
@@ -1933,6 +1934,14 @@ def lstsq(a, b, rcond=-1):
+    #    not have this bug:
+    #      http://icl.cs.utk.edu/lapack-forum/archives/lapack/msg00899.html
+    #    Lapack_lite does have that bug...
     nlvl = max( 0, int( math.log( float(min(m, n))/2. ) ) + 1 )
```
According to LAPACK, this should be:

> SMLSIZ is returned by ILAENV and is equal to the maximum
> size of the subproblems at the bottom of the computation
> tree (usually about 25), and
>    NLVL = MAX( 0, INT( **LOG_2**( MIN( M,N )/(**SMLSIZ**+1) ) ) + 1 )

For whatever reason, we seem to have decided that `log(2) = SMLSIZ = 1`, which seems best described as "false". Either way, this is a can of worms, and not one that I think this PR should be opening.

seberg replied:
Should we open an issue about this? Is there some example for when this goes bad? (I really hate bugs that might silently create incorrect results, and if this is the case, we should give it some priority)

eric-wieser replied:
I have a patch in the works that uses lapack's internal mechanism to calculate this correctly. Unfortunately, there's a bug in the modified version of lapack bundled with numpy that requires us to regenerate the lapack c code from a newer-but-not-so-new-to-break-f2c release of lapack  (#8376). The next step of fixing this is to actually get the generator running again ( #8381 ).

seberg replied:
Great, I have no idea of these things, was just worried this might be forgotten :)

eric-wieser replied:
Guess it would do no harm to open an issue. My working tree was branching way too much for a bunch of problems discovered while trying to fix the problems, and I think the best call is to sit tight and wait for PR merge/rejection, rather than increasing the amount of rebases I need to do each time!

eric-wieser committed:
BUG: prevent np.linalg.eigh ValueError when given a 0x0 array

Throw LinAlgError instead with an appropriate message

eric-wieser commented:
@pv: Squashed, strimmed-in-scope, and passing. Let me know if anything is still unclear

charris commented on numpy/linalg/tests/test_linalg.py:
```diff
@@ -61,30 +61,44 @@ def get_rtol(dtype):
 class LinalgCase(object):
 
-    def __init__(self, name, a, b, exception_cls=None):
+    def __init__(self, name, a, b, flags=frozenset()):
```
The class should be documented, at least for the constructor. This is especially so now with the rather generic `flags`.

charris replied:
Why not a simple list instead of a `frozenset`? I don't see that an immutable set really matters here.

charris commented on numpy/linalg/tests/test_linalg.py:
```diff
@@ -320,43 +348,87 @@ def _check_cases(func, cases):
+    def test_nonsq_cases(self):
+        _check_cases(self.do,
+            require=CaseFlags.none,
+            exclude=CaseFlags.generalized | CaseFlags.square | CaseFlags.hermitian | CaseFlags.empty)
```
See above. I think these would all be clearer as a list of strings and take up less room to boot.

charris replied:
My counter argument would be that it isn't important ;) If you were writing fly-by-wire software that had to be 100% correct at all times, then using an uncommon python feature would be justified. In fact, seeing such usage makes me stop and think and track down why it is important. But here I don't think it matters and I've burned up a couple of precious grey cells to no purpose. The simplest approach that works is what I would chose in this case because it makes the code that much more accessible, and likely more error free as it is easier to check.

charris commented on numpy/linalg/tests/test_linalg.py:
```diff
@@ -729,7 +831,10 @@ def test_square(self):
-    def do(self, a, b):
+    def do(self, a, b, flags):
+        if flags & CaseFlags.empty:
+            assert_raises(LinAlgError, linalg.eigvalsh, a, 'L')
```
So checking for an error is the default (no flags). That is somewhat unintuitive.

eric-wieser replied:
Darn, It turns out `empty` is a confusing name. Errors only occur when an array has `size=0`.

charris commented:
I think the tests mods are overly complicated, which makes them more difficult to follow, which makes them harder to maintain.

Note that the commits are not squashed.

eric-wieser committed:
fixup! TST: Refactor all the test case lists

eric-wieser commented:
@charris I've tried to simplify things now. See the fixup commit above for the diff (I figured it'd be easier to review before applying it)

charris commented on numpy/linalg/tests/test_linalg.py:
```diff
@@ -201,21 +233,16 @@ def __repr__(self):
 #
 # Gufunc test cases
 #
+def _make_generalized_cases():
```
Is there really an advantage in having a function with no arguments that operates on a global and is called once? I think the original is cleaner and easier to follow.

eric-wieser replied:
Avoiding global namespace pollution - stops anything accidentally using these local variables in a test. Also handy for code folding, sort of

charris replied:
Fair enough.

charris commented on numpy/linalg/tests/test_linalg.py:
```diff
@@ -289,29 +319,35 @@ def _stride_comb_iter(x):
-            GENERALIZED_NONSQUARE_CASES,
-            GENERALIZED_HERMITIAN_CASES):
-
+def _make_strided_cases():
```
See above. I don't see the use in making this a function.

charris commented on numpy/linalg/tests/test_linalg.py:
```diff
@@ -289,29 +319,35 @@ def _stride_comb_iter(x):
+    """
+    for case in CASES:
+        # filter by require and exclude
+        if case.tags & require != require:
```
Hmm, this is a bit clumsy for a lookup, although I suspect it would require a redesign to improve it. Probably not worth the effort.

charris commented:
Couple of style comments. I note this hasn't been squashed, oversight or do you want help?

eric-wieser commented:
@charris: I partially squashed it, but out of preference of trying to keep logically separate changes, not into one commit.

I don't need help performing the squash, but would perhaps like some advice on what level to squash things to for numpy

charris commented:
I'd probably have squashed all the `TST` commits, then added the `BUG` fixes on top of that. However, as long as none of the commits errors out it is not a big problem. However, not all of the commit messages are up to snuff. For instance
``` 
TST: Correct test cases to actually make sense

These testcases were never used in the first place, due to a typo. This makes
their dimensions match the order of the other test cases, even though those
also did not run
```
Doesn't impart much information because it requires context, it is basically "fix some stuff", whatever "stuff" is. A virtuoso -- Al Viro of Linux fame -- would probably redo the series of commits into a logical rather than temporal series, but we aren't that demanding...

Anyway, I'm tempted to do a squash commit on this.

charris commented:
In it goes, thanks Eric.

eric-wieser commented:
Glad I didn't start rebasing those logically then!

eric-wieser commented:
Thanks, I'll rebase #8649 onto master next time I work on it (and after #8651), so it gets the more thorough testing

eric-wieser commented:
Oops, looks like we left a !fixup in there
